### ISSUE TITLE
Runtime exception when injector is configured literaly before the injector class

### ISSUE THREAD

**Issue by NumberFourAG** [NumberFour/n4js#28]
IDEBUG-365
 //edit
problem is {{C}} is used before it had value defined. this is not problem of DI, but of our compiled code in general, e.g:

```
console.log("@1", C);// @1 undefined
class C {}
console.log("@2", C);// @2 function C() {}
```

//orginal
The below snippet throws runtime exception:

```
var injector = ((C as Object)['$diInjector'] as Function)();
var C c = injector.createnew(C) as C;

@Injector
class C {

    @Inject
    private S s;

    public getSingleton() { return this.s; }

}

class S {

}

console.log(c.getSingleton() === c.getSingleton)
```

But works fine if I change the injector setup order:

```
@Injector
class C {

    @Inject
    private S s;

    public getSingleton() { return this.s; }

}

class S {

}

var injector = ((C as Object)['$diInjector'] as Function)();
var C c = injector.createnew(C) as C;

console.log(c.getSingleton() === c.getSingleton)
```

Configuration is supposed to be derived from annotations, and generated automatically (should be done as part of IDE-1275). Manual configuration is just workaround before we implement it. Once we have it manual configuration should be removed and this bug is invalid.

Even with build in type this bug occurs, e.g. 

```
var injector = N4Injector.of(C);//using C before it is defined;
@GenerateInjector
class C {
    @Inject
    private S s;
    public getSingleton() { return this.s; }
}
class S {}
var C c = injector.create(C);
console.log(c.getSingleton() === c.getSingleton)
```

problem is {{C}} is used before it had value defined. this is not problem of DI, but of our compiled code in general, e.g:

```
console.log("@1", C);// @1 undefined
class C {}
console.log("@2", C);// @2 function C() {}
```

**Comment by mor-n4** [comment]
updated above example to the new DI API:

```
import {N4Injector} from "n4js/lang/N4Injector";

var injector = N4Injector.of(C);
var c: C = injector.create(C);

@GenerateInjector
class C {

    @Inject
    private s: S;

    public getSingleton() { return this.s; }
}

class S {
}

console.log(c.getSingleton() === c.getSingleton)
```

**Comment by mor-n4** [comment]
as someone already noted in the "edit" section at the beginning of the issue description, this is not really a problem of DI, but of how we set up classes and interfaces with `$makeClass` etc.

This problem has been solved partially (hoisting of classifier declaration). Therefore, the following example from the task description

```
console.log("@1", C);// @1 undefined
class C {}
console.log("@2", C);// @2 function C() {}
```

now actually prints out:

```
@1 function C() {}
@2 function C() {}
```

However, this is solved only partially. Some information on classifiers is still undefined above the declaration:

```
console.log(new C().field); // ok-ish (prints 'hello')
console.log(C.fieldStatic); // error (prints 'undefined')
console.log(C.n4type); // error (prints 'undefined')

class C {
    field = "hello";
    static fieldStatic = "world!";
}
```

HOWEVER: since ES6 class declarations aren't hoisted and we want to bring N4JS classes closer to ES6 classes, we might also just have to live with the problem reported in this issue.

Anyways, we have tasks for refining our classes w.r.t. ES6 classes, so this bug here seems like a duplicate.
